### PR TITLE
Main menu overhaul (Part 1):  Mouse controls + Vertical submenus

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

### Purpose of change
- Contributes to #57098

### Describe the solution

This is just a first pass introducing some QoL improvements to the main menu:
- Allow mouse controls for menu and submenu buttons (SDL only)
- Standardize on vertical secondary menus
- Use `uilist`s for tertiary menus

<hr>

#### Mouse controls

Primary, secondary and tertiary menu options are mouse accessible on SDL builds (non-curses). Moving the mouse also highlights primary and secondary menu options:

https://user-images.githubusercontent.com/12537966/168766532-1fbc17a2-cace-40a1-969d-1fb784a0be54.mp4

Left-clicking on a menu item selects it and confirms the option (where applicable). Mouse movement does nothing on tertiary menus since they rely on `uilist` for input handling.

The mouse wheel scrolls through vertical submenu options, as well as scrollable text like the MOTD/Credits windows.

<hr>

#### Vertical submenus

Secondary menus use a common vertical layout:

vert_submenu

A couple of extra minor changes:
- Submenus (where applicable) are automatically displayed when switching main menu options
  - Menu items are navigated with `LEFT/BACKTAB` and `RIGHT/TAB`
  - Submenu items are navigated with `UP/PAGEUP` and `DOWN/PAGEDOWN`
  - Confirming a submenu option is done with `ENTER`
- Hints for "New Character" options are listed on the line below the main menu options (in yellow)
- Submenu hotkeys were revised to avoid conflicts with the main menu (since both are accessible at the same time)

<hr>

### Describe alternatives you've considered
This is just a first pass addressing some of the items from #57098 (mostly limited by the size of these changes). Future changes will address the problems with the Settings and World Options menus to round off the issue.

### Testing
See clips above. Manually tested each menu and submenu item to check that they behave the same as before.

Also tested with the minimal resolution, just to make sure each submenu is accessible:

win_small

Also tested the curses build:

win_curses

### Additional context
I took some liberties with some of the design elements, but I think it works pretty well.

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
Awesome!

Is it possible to highlight element that you mouse-clicked on first, and then start its function, not the way around, as it is now?

**Comment by dseguin** [comment]
It's possible, for sure. I'm not sure if that's the way Erk would want it, since it would add extra clicks (and on touchscreen controls you would have to tap the option twice, but tapping twice also acts as `ESC`)

C.C.: @I-am-Erk

**Comment by Night-Pryanik** [comment]
No, you misunderstood me. No need to double click: just highlight clicked option and then start its function right after that - all with one click.

**Comment by dseguin** [comment]
Oh I see, my mistake :P

The short answer is "yes", but it would add a lot of complexity. (drawing happens before input, so the action would need to be deferred for a frame, and the next input would be queried on the frame after that)

File: src/main_menu.cpp [hunk 1]
```diff
@@ -92,34 +111,155 @@ static int utf8_width_notags( const char *s )
...
-void main_menu::print_menu( const catacurses::window &w_open, int iSel, const point &offset )
+void main_menu::display_sub_menu( int sel, const point &bottom_left, int sel_line )
 {
+    main_menu_sub_button_map.clear();
+    std::vector<std::string> sub_opts;
+    int xlen = 0;
+    main_menu_opts sel_o = static_cast<main_menu_opts>( sel );
+    switch( sel_o ) {
+        case main_menu_opts::CREDITS:
+            display_text( mmenu_credits, "Credits", sel_line );
```
Comment (Night-Pryanik) [inline src/main_menu.cpp:173]: Is "Credits" here translated?

File: src/main_menu.cpp [hunk 2]
```diff
@@ -283,6 +432,32 @@ void main_menu::init_strings()
...
     vMenuItems.emplace_back( pgettext( "Main Menu", "<Q|q>uit" ) );
 
+    // new game menu items
+    vNewGameSubItems.clear();
+    vNewGameSubItems.emplace_back( pgettext( "Main Menu|New Game", "C<u|U>stom Character" ) );
+    vNewGameSubItems.emplace_back( pgettext( "Main Menu|New Game", "<P|p>reset Character" ) );
+    vNewGameSubItems.emplace_back( pgettext( "Main Menu|New Game", "<R|r>andom Character" ) );
+    if( !MAP_SHARING::isSharing() ) { // "Play Now" function doesn't play well together with shared maps
+        vNewGameSubItems.emplace_back( pgettext( "Main Menu|New Game",
+                                       "Play Now!  (<F|f>ixed Scenario)" ) );
```
Comment (Night-Pryanik) [inline src/main_menu.cpp:442]: Maybe change "Fixed Scenario" to "Default Scenario"?

File: src/main_menu.cpp [hunk 3]
```diff
@@ -597,268 +792,97 @@ bool main_menu::opening_screen()
...
-        ui_manager::redraw();
-        if( layer == 2 && sel1 == 1 ) {
-            // Then choose custom character, random character, preset, etc
-            if( MAP_SHARING::isSharing() &&
-                world_generator->all_worldnames().empty() ) { //don't show anything when there are no worlds (will not work if there are special maps)
-                layer = 1;
-                sel1 = 1;
-                continue;
+        while( 1 ) {
+            uilist mmenu( _( "Choose a preset character template." ), {} );
```
Comment (Night-Pryanik) [inline src/main_menu.cpp:858]:
```suggestion
            uilist mmenu( _( "Choose a preset character template" ), {} );
```

File: src/main_menu.cpp (hunk 1 above)
Comment (dseguin) [reply src/main_menu.cpp:173]: Nope, but it's strange that it's not translated in master either:

https://github.com/CleverRaven/Cataclysm-DDA/blob/556afb9bf7cceca4db5a43c0dcac3a10fd2439bc/src/main_menu.cpp#L445-L449

**Comment by leemuar** [comment]
> No, you misunderstood me. No need to double click: just highlight clicked option and then start its function right after that - all with one click.

I think it might be better to highlight on "mouse over" and to activate on click. I have no idea how to implement Mouse Over events on menu items though...

**Comment by dseguin** [comment]
> I think it might be better to highlight on "mouse over" and to activate on click.

Turns out this was pretty easy:

https://user-images.githubusercontent.com/12537966/168766532-1fbc17a2-cace-40a1-969d-1fb784a0be54.mp4

**Comment by leemuar** [comment]
> Turns out this was pretty easy:

This is absolutely awesome addition, thanks!

**Comment by ZeroInternalReflection** [comment]
This definitely fixes some of the quirks of the main menu.  One note on the mouseover behaviour, however, is that the credits/MOTD displays get a bit tricky on small screens.  You can get kicked out of the screen by moving your mouse such that it lands on a menu that you can't see (The video is in Ukrainian since I also wanted to test whether everything works when the menu options get bumped to a second line, and it appears to):
https://user-images.githubusercontent.com/89038572/168846661-63a935d8-86b6-4c1c-81dd-46d540c32dac.mp4

**Comment by Inglonias** [comment]
what

**Comment by dseguin** [comment]
I'm not sure how to address this. I see 2 solutions, both involve adding a special case for TERMX size:
- Prevent mouse functions when MOTD/Credits are selected
- Require confirming the MOTD/Credits option to view the text in a new window

**Comment by ZeroInternalReflection** [comment]
> I'm not sure how to address this. I see 2 solutions, both involve adding a special case for TERMX size:
> 
>     * Prevent mouse functions when MOTD/Credits are selected
> 
>     * Require confirming the MOTD/Credits option to view the text in a new window

Is there a particular reason that display_text() is currently fixed at FULL_SCREEN_HEIGHT/WIDTH?  Rather than having a slightly different mouse interface for the credits/MOTD, I'd favour just capping the window height in display_text().  I feel consistency in interface behaviour is more important than maximizing the size of the credits/MOTD panes.  This test video uses a fixed allowance of 4 rows for tips/menu/etc.  This should ensure the menu you're mousing over is always visible, no matter what screen size/language you're using.

https://user-images.githubusercontent.com/89038572/168946776-d4213347-cf00-41af-82b6-936f459f0ab1.mp4

**Comment by dseguin** [comment]
Cheers, that's a much better solution. Not sure why I didn't consider it, haha.

**Comment by leemuar** [comment]
I think it would be better to not activate "main" menus (MOTD, New Game, Load, World, Special, Settings, Help, Credits, Quit) by mouse hovering, it should require a click. It should be highlighted, but not activated. 
Submenus however should act just like it is nor

P.S. again, thanks for this change, it really made the game much more "reactive" and responsive to the player

**Comment by dseguin** [comment]
About this:

> not activate "main" menus [...] by mouse hovering, it should require a click

I tested hiding the submenus when mousing over:



In my opinion, this creates a bit more complexity in the controls, since cycling the options using keys behaves differently than mousing over. It also goes against this point from #57098:

> Submenu options (such as the new world/select a world dialogue above) automatically become visible when you highlight the main option ([World] in this case). You don't press enter or space to open a submenu. This reduces a lot of keypresses and menu navigation confusion at no cost, since because of item 3 below, if this isn't the menu you want, you just keep on scrolling.

**Comment by leemuar** [comment]
> In my opinion, this creates a bit more complexity in the controls, since cycling the options using keys behaves differently than mousing over. It also goes against this point from https://github.com/CleverRaven/Cataclysm-DDA/issues/57098:

These are good argument, I haven't thought about that. Thanks for making it clear for me!

**Comment by dseguin** [comment]
> How does it respond to a touch screen?

I don't have the hardware to test this, but from looking around it looks like touch gestures are mapped to keys. So tapping once is just a `.` input, which means "confirm".

I don't think it's possible to check for touchscreen tap coordinates without doing some voodoo magic (or implementing a native Android UI like in #53132). Brett also pointed out that the buttons might be too small to press accurately on a touch screen

File: src/main_menu.cpp [hunk 4]
```diff
@@ -411,15 +589,19 @@ bool main_menu::opening_screen()
 
     load_char_templates();
 
-    ctxt.register_cardinal();
+    ctxt.register_directions();
```
Comment (leemuar) [inline src/main_menu.cpp:592]: Why directions? I see no use of LEFTUP, LEFTDOWN, RIGHTUP or RIGHTDOWN action

Comment (dseguin) [reply src/main_menu.cpp:592]: Yep, that was a slip-up from testing. Cheers.

```suggestion
    ctxt.register_cardinal();
```

**Comment by Night-Pryanik** [comment]
Should this resolve #53699?

**Comment by dseguin** [comment]
> Should this resolve https://github.com/CleverRaven/Cataclysm-DDA/issues/53699?

Nope, this PR doesn't affect the character creation menu

**Pull request merged**